bundles/io/org.eclipse.smarthome.io.rest.log/src/main/java/org/eclipse/smarthome/io/rest/log/internal/LogHandler.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.rest.log.internal;
+
+import java.net.URL;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import org.eclipse.smarthome.io.rest.RESTResource;
+import org.eclipse.smarthome.io.rest.log.LogConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+
+@Path("/log")
+@Produces(MediaType.APPLICATION_JSON)
+public class LogHandler implements RESTResource {
+
+    private final Logger logger = LoggerFactory.getLogger(LogHandler.class);
+
+    /**
+     * Rolling array to store the last LOG_BUFFER_LIMIT messages. Those can be fetched e.g. by a
+     * diagnostic UI to display errors of other clients, where e.g. the logs are not easily accessible.
+     */
+    private final ArrayDeque<LogMessage> LOG_BUFFER = new ArrayDeque<>(LogConstants.LOG_BUFFER_LIMIT);
```

Comment:
ArrayDeque is not thread save. If multiple threads are logging, you would risk a ConcurrentModificationException. Please change it to e.g. ConcurrentLinkedDeque.